### ISSUE TITLE
Net::SSH issues when exploiting older hosts.

### ISSUE THREAD

**Issue by darrenmartyn** [rapid7/metasploit-framework#16138]
I noticed this issue when testing the tectia_passwd_changereq module, but it probably affects every module using Net::SSH.

TL;DR Net::SSH prefers to use only modern, good algorithms for SSH, wheras vulnerable boxes tend to use crap, insecure algorithms.

```
(unix/ssh/tectia_passwd_changereq) > run

[-] nope:22 - nope:22 SSH Error: Net::SSH::Exception : could not settle on kex algorithm
Server kex preferences: diffie-hellman-group1-sha1,extension1-sha1@ssh.com
Client kex preferences: ecdh-sha2-nistp521,ecdh-sha2-nistp384,ecdh-sha2-nistp256,diffie-hellman-group-exchange-sha256,diffie-hellman-group14-sha1

```

## Expected behavior

SSH connects and does stuff.

## Actual behaviour

The SSH client (Metasploit, specifically, Net::SSH) can't agree with the server on algorithms to use. 

## Suggested fix

Given the usecase, Metasploit should patch/somehow fix its version of Net::SSH to permit it to use insecure algorithms, so insecure computers can be exploited.

**Comment by adfoster-r7** [comment]
@darrenmartyn Thanks for the issue, do you have an example of an older box that you've tried this against?

A quick look shows that they they removed `diffie-hellman-group1-sha1,extension1-sha1@ssh.com` kex a few years back from Net::SSH as part of a "Prefer more modern cryptographic algorithms' commit https://github.com/net-ssh/net-ssh/commit/a38ab386f060e176ddb58b92c4eec9170b42dd2e#diff-d3f3cd24513081ec6736e7516607d8de54cb329fe6319a2c3fabe4edd67b4393L32

It looks like there's a configuration option to bring it back the older functionality, although it looks like the next major version will remove those algorithms entirely:

> Net::SSH 6.0 disables by default the usage of weak algorithms. We strongly recommend that you install a servers's version that supports the latest algorithms.
>
> It is possible to return to the previous behavior by adding the option : append_all_supported_algorithms: true
> 
> Insecure algoritms will definitely be removed in Net::SSH 7.*.e`
> https://github.com/net-ssh/net-ssh/blame/0150d054f0cd0beacd4ba1000c6df6d8636a2c18/README.md#L34

**Comment by darrenmartyn** [comment]
The system I was testing against is internal, it is a legacy Unix (AIX, specifically) system with an ancient version of the Tectia SSH server installed. I don't have any public facing systems that I can test the module against actively. 

From another research project of mine (a personal instance of the [passive-ssh](https://github.com/D4-project/passive-ssh) scanner) I can also find a fairly substantial number of servers on the global internet that also only support older (insecure) algorithms, which leads me to believe that this issue probably is going to impact more than just this one module/usecase.

**Comment by adfoster-r7** [comment]
Thanks! I've ran into this issue within labs too, was just hoping to save time setting up a test environment if were testing against a known hackthebox/vulnhub/etc machine 😄

**Comment by heyder** [comment]
Shouldn't we close this issue?

**Comment by bcoles** [comment]
Fixed in #16318.
